### PR TITLE
Pretty Print Generators and Effects

### PR DESCRIPTION
**Pull request by cblappert**
In an effort to make debugging easier, add utilities that help pretty-print generators and effects with all the important fields (and none of the internal ones that you see in a debugger). Depth of recursion of printer is customizable.

Currently doesn't pass Flow because I have a helper function that takes an Object and returns an Object. Not sure how to make that type more specific.

Open to feedback on ways to change the output to be more readable/useful.

Output looks like this for a simple set of effects:
```
{
  result: [ReturnCompletion value [Abstract 1735003607742184]],
  generator: {
    _entries: {
      index: 39,
      declared: [Abstract 1735003607742184],
      args: []
    },
    id: 35,
    _name: evaluateNodeForEffects,
    pathConditions: [[Abstract 1735003607742177], [Abstract 1735003607742183]]
  },
  canBeApplied: true,
  _id: 33
}
```

### CODE REVIEW

File: src/realm.js [hunk 1]
```diff
@@ -111,6 +112,15 @@ export class Effects {
   createdObjects: CreatedObjects;
   canBeApplied: boolean;
   _id: number;
+
+  toDisplayString() {
```
Comment (hermanventer) [inline src/realm.js]: Please always add a return type annotation.

File: src/types.js [hunk 2]
```diff
@@ -1055,6 +1055,10 @@ export type UtilsType = {|
   typeToString: (typeof Value) => void | string,
   getTypeFromName: string => void | typeof Value,
   describeValue: Value => string,
+  // eslint-disable-next-line flowtype/no-weak-types
+  jsonToDisplayString: <T: { toDisplayJson(number): Object }>(T) => string,
```
Comment (hermanventer) [inline src/types.js]: The Flow guide suggests that you use {} rather than Object.

File: src/utils.js [hunk 3]
```diff
@@ -106,3 +106,41 @@ export function describeValue(value: Value): string {
...
+// eslint-disable-next-line flowtype/no-weak-types
+export function verboseToDisplayJson(obj: Object, depth: number): Object {
+  let result = {};
+  function valueOfProp(prop) {
+    if (typeof prop === "function") return undefined;
+    if (Array.isArray(prop)) {
+      // Try to return a 1-line string if possible
+      if (prop.length === 0) return "[]";
+      let valuesArray = prop.map(x => valueOfProp(x));
+      if (valuesArray.length === 1) return valuesArray[0];
```
Comment (hermanventer) [inline src/utils.js]: It appears that [1] will display as just 1. Is that what you want?

File: src/utils.js [hunk 4]
```diff
@@ -106,3 +106,41 @@ export function describeValue(value: Value): string {
...
+      if (valuesArray.length < 5) {
+        let string =
+          "[" + valuesArray.reduce((acc, x) => `${acc}, ${x instanceof Object ? JSON.stringify(x) : x}`) + "]";
+        string = string.replace(/\"/g, "");
+        if (string.length < 60) return string;
+      }
+      return valuesArray;
+    }
+    if (prop instanceof Set || prop instanceof Map) return prop.size;
```
Comment (hermanventer) [inline src/utils.js]: I think you should include the type prop here.

File: src/realm.js [hunk 5]
```diff
@@ -900,6 +911,7 @@ export class Realm {
         // Return the captured state changes and evaluation result
         if (c instanceof Value) c = new SimpleNormalCompletion(c);
         result = new Effects(c, astGenerator, astBindings, astProperties, astCreatedObjects);
+        console.log(result.toDisplayString());
```
Comment (NTillmann) [inline src/realm.js]: left over from debugging?

**Comment by calebmer** [comment]
For generators, how useful do you think it would be to `toString` their `buildNode`? So we could see the source of the function. You can kind of get an idea from `_entries` but it still requires some thought.

**Comment by cblappert** [comment]
@trueadm is refactoring how buildNodes work so that every buildNode will have a different type/kind key. Once he adds that, I can add that to the logging.

**Pull request closed without merging**